### PR TITLE
WIP: JSON Schema Validation, Part 1

### PR DESCRIPTION
**Pull request by mrkaye97**
WIP. Opening this PR to refer back to in #5208 

## What changes are proposed in this pull request?

This PR is a first pass at adding JSON validation against a predetermined schema, in an effort to make error handling more friendly for users. The end goal is to return `HTTP 400` for bad parameters instead of `500`, which happens currently (at least for most endpoints)

## How is this patch tested?

WIP -- Not tested yet. What is the best way to test changes like this?

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by mrkaye97** [comment]
See [this comment](https://github.com/mlflow/mlflow/issues/5208#issuecomment-1030858836) for some context on the issue I'm grappling with here. I may just be missing something about `parse_dict` since I'm not familiar with it, but at least afaict it looks like we might need to do extra type conversion or JSON parsing for this to work.

File: mlflow/server/handlers.py [hunk 1]
```diff
@@ -323,6 +345,31 @@ def _not_implemented():
...
+        {
+          "title": "experiments/create",
+          "properties": {
+            "name": {
+              "type": "string",
+              "description": "Experiment name."
+            },
+            "artifact_location": {
+              "type": "string",
+              "description": "Location where all artifacts for the experiment are stored. If not provided, the remote server will select an appropriate default."
```
Comment (mrkaye97) [inline mlflow/server/handlers.py:359]: Copied from the REST API doc

**Comment by mrkaye97** [comment]
@dbczumar Moving discussion over here.
1. What is the best way to test these changes? Is CI passing (waiting on that currently) enough to be confident in them? Are there tests living somewhere that I should be running on local before pushing? 
2. [This](https://github.com/mlflow/mlflow/pull/5350/files#diff-5c10a4e2ca47745f06fa9e7201087acfc102849756cb8d85e774a5ac468cb037R225-R229) seems like a reasonable approach to me, for now at least. It'll let me keep developing in smaller batches. WDYT?
3. [Is this the proper way to raise a 400 for bad params](https://github.com/mlflow/mlflow/pull/5350/files#diff-5c10a4e2ca47745f06fa9e7201087acfc102849756cb8d85e774a5ac468cb037R239-R243)?
4. I'm not sure how I feel about [including the schema in the methods](https://github.com/mlflow/mlflow/pull/5350/files#diff-5c10a4e2ca47745f06fa9e7201087acfc102849756cb8d85e774a5ac468cb037R349-R371). I think it'd probably be better to define a big object somewhere with _all_ of the schemas, and then look up the one we need instead of defining it inside of each method. WDYT?

**Comment by dbczumar** [comment]
> @dbczumar Moving discussion over here.
> 
> 1. What is the best way to test these changes? Is CI passing (waiting on that currently) enough to be confident in them? Are there tests living somewhere that I should be running on local before pushing?
> 2. [This](https://github.com/mlflow/mlflow/pull/5350/files#diff-5c10a4e2ca47745f06fa9e7201087acfc102849756cb8d85e774a5ac468cb037R225-R229) seems like a reasonable approach to me, for now at least. It'll let me keep developing in smaller batches. WDYT?
> 3. [Is this the proper way to raise a 400 for bad params](https://github.com/mlflow/mlflow/pull/5350/files#diff-5c10a4e2ca47745f06fa9e7201087acfc102849756cb8d85e774a5ac468cb037R239-R243)?
> 4. I'm not sure how I feel about [including the schema in the methods](https://github.com/mlflow/mlflow/pull/5350/files#diff-5c10a4e2ca47745f06fa9e7201087acfc102849756cb8d85e774a5ac468cb037R349-R371). I think it'd probably be better to define a big object somewhere with _all_ of the schemas, and then look up the one we need instead of defining it inside of each method. WDYT?

@mrkaye97 Great questions!

1. In addition to CI on this PR, running https://github.com/mlflow/mlflow/blob/master/tests/tracking/test_rest_tracking.py via `cd tests/tracking && pytest test_rest_tracking.py` with `pytest==6.2.5` will provide a good sense of whether or not things are working.

2. Sounds great

3. We recommend raising `MlflowException` with error code `INVALID_PARAMETER_VALUE`. Here's an example: https://github.com/mlflow/mlflow/blob/9c70732d4f380ded5a31a2ca76686306956189d8/mlflow/pyfunc/__init__.py#L819-L823

4. Do we need to store the full schema? Perhaps it would be sufficient to have the validators access attributes of `request_message` and assert on them? This logic can be specific to each validator.

**Comment by mrkaye97** [comment]
Great, thanks for the feedback. That all makes sense. On 4), are you thinking something [like this](https://github.com/mlflow/mlflow/pull/5350/files#diff-5c10a4e2ca47745f06fa9e7201087acfc102849756cb8d85e774a5ac468cb037R283-R287) and [this](https://github.com/mlflow/mlflow/pull/5350/files#diff-5c10a4e2ca47745f06fa9e7201087acfc102849756cb8d85e774a5ac468cb037R222-R224) where I've just factored out some schema logic into a dictionary? This seems problematic to me because it's too automagical, but I'm not 100% sure if I was understanding your suggestion correctly.

Another thing that's probably cleaner would be to do something like this:

```python
def _name_schema(**kwargs):
  base = {"type": "string"}
  for k, v in **kwargs:
    base[k] = v
```

(excuse ugly python -- not much of a pythoner 🙂)

This way we would be able to pass additional things, like `"required": true` or `"maxItems": 10` or something, and then could validate similarly:

```python
Draft202012Validator(_name_schema()).is_valid(request_json["name"])
```

I'm sure this could get cleaned up or made a little more Pythonic, but it seems like a reasonable strategy to me

**Comment by dbczumar** [comment]
> 

@mrkaye97 This looks fantastic (https://github.com/mlflow/mlflow/pull/5350/files#diff-5c10a4e2ca47745f06fa9e7201087acfc102849756cb8d85e774a5ac468cb037R283-R287 / https://github.com/mlflow/mlflow/pull/5350/files#diff-5c10a4e2ca47745f06fa9e7201087acfc102849756cb8d85e774a5ac468cb037R222-R224). I don't have a strong preference either way between the automagical approach and the second one you've outlined. Both seem to translate a dictionary of field names and types into a validation routine, right?

**Comment by dbczumar** [comment]
Hi @mrkaye97 , any updates here? Let me know if there's anything we can help with.

**Comment by mrkaye97** [comment]
Workin on it! Been busy at the day job for the past week or so. I'll plan on taking another look tomorrow or Thursday and report back, but I think I've got a pretty good idea of how to design an initial pass at this point. 

Separately, I'm curious about the test failures I'm seeing in CI. For instance, in the R tests it seems like there are a lot of flaky tests failing after timing out waiting for a server to spin up. I'm mostly just wondering is this is the usual state of affairs for these tests -- was seeing similar things for the Python tests too, where things were failing for completely unrelated reasons to any of these changes that were unpredictable.

**Comment by mrkaye97** [comment]
Closing in favor of #5458

**Pull request closed without merging**